File: src/ormai/adapters/django/adapter.py
"""
Django ORM adapter implementation.

Provides query compilation and execution for Django models.
"""

from collections.abc import Callable
from typing import Any, TypeVar

from django.db import models, transaction
from django.db.models import Q, F, Count, Sum, Min, Max, Avg

from ormai.adapters.base import CompiledQuery, OrmAdapter
from ormai.adapters.django.introspection import DjangoIntrospector
from ormai.core.context import RunContext
from ormai.core.dsl import (
    AggregateRequest,
    AggregateResult,
    CreateRequest,
    CreateResult,
    DeleteRequest,
    DeleteResult,
    FilterClause,
    GetRequest,
    GetResult,
    QueryRequest,
    QueryResult,
    UpdateRequest,
    UpdateResult,
)
from ormai.core.errors import ModelNotAllowedError, NotFoundError
from ormai.core.types import SchemaMetadata
from ormai.policy.engine import PolicyEngine
from ormai.policy.models import Policy
from ormai.policy.scoping import ScopeInjector

T = TypeVar("T")


class DjangoAdapter(OrmAdapter):
    """
    OrmAI adapter for Django ORM.

    Supports Django models with policy-governed queries.

    Usage:
        from django.apps import apps
        from ormai.adapters.django import DjangoAdapter

        adapter = DjangoAdapter(apps.get_app_config('myapp'))
    """

    def __init__(
        self,
        app_config: Any = None,
        models: list[type[models.Model]] | None = None,
    ) -> None:
        """
        Initialize the Django adapter.

        Args:
            app_config: Django AppConfig for the app (optional)
            models: Explicit list of models to include (optional)
        """
        self.app_config = app_config
        self._models = models or []
        self._introspector = DjangoIntrospector(app_config, models)
        self._model_map: dict[str, type[models.Model]] = {}

    async def introspect(self) -> SchemaMetadata:
        """Introspect Django models."""
        schema = self._introspector.introspect()
        # Build model map for quick lookup
        self._model_map = self._introspector.get_model_map()
        return schema

    def sync_introspect(self) -> SchemaMetadata:
        """Synchronous introspection."""
        schema = self._introspector.introspect()
        self._model_map = self._introspector.get_model_map()
        return schema

    def _get_model(self, model_name: str) -> type[models.Model]:
        """Get the Django model class by name."""
        if model_name not in self._model_map:
            raise ModelNotAllowedError(
                model_name,
                allowed_models=list(self._model_map.keys()),
            )
        return self._model_map[model_name]

    def compile_query(
        self,
        request: QueryRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile a query request into a Django QuerySet."""
        engine = PolicyEngine(policy, schema)
        model_policy = engine.validate_model_access(request.model)

        # Get Django model class
        model_class = self._get_model(request.model)

        # Start with base queryset
        queryset = model_class.objects.all()

        # Inject scope filters
        injector = ScopeInjector(policy)
        scope_filters = injector.get_scope_filters(request.model, ctx)

        # Build Q objects for filters
        q_objects = []
        for f in scope_filters:
            q_objects.append(self._filter_to_q(f))

        # Apply request filters
        if request.where:
            for f in request.where:
                q_objects.append(self._filter_to_q(f))

        # Apply all filters
        for q in q_objects:
            queryset = queryset.filter(q)

        # Apply field selection
        select_fields = engine.filter_select_fields(
            request.model,
            request.select,
            model_policy,
        )

        if select_fields:
            queryset = queryset.values(*select_fields)

        # Apply ordering
        if request.order_by:
            order_fields = []
            for o in request.order_by:
                field = o.field
                if o.direction == "desc":
                    field = f"-{field}"
                order_fields.append(field)
            queryset = queryset.order_by(*order_fields)

        # Apply pagination
        if request.take:
            queryset = queryset[:request.take]

        return CompiledQuery(
            query=queryset,
            request=request,
            select_fields=select_fields,
            injected_filters=scope_filters,
            policy_decisions=[f"model_allowed:{request.model}"],
        )

    def compile_get(
        self,
        request: GetRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile a get-by-id request."""
        engine = PolicyEngine(policy, schema)
        model_policy = engine.validate_model_access(request.model)

        model_class = self._get_model(request.model)
        queryset = model_class.objects.all()

        # Inject scope filters
        injector = ScopeInjector(policy)
        scope_filters = injector.get_scope_filters(request.model, ctx)

        for f in scope_filters:
            queryset = queryset.filter(self._filter_to_q(f))

        # Filter by primary key
        pk_field = self._get_pk_field(model_class)
        queryset = queryset.filter(**{pk_field: request.id})

        # Apply field selection
        select_fields = engine.filter_select_fields(
            request.model,
            request.select,
            model_policy,
        )

        if select_fields:
            queryset = queryset.values(*select_fields)

        return CompiledQuery(
            query=queryset,
            request=request,
            select_fields=select_fields,
            injected_filters=scope_filters,
        )

    def compile_aggregate(
        self,
        request: AggregateRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile an aggregation request."""
        engine = PolicyEngine(policy, schema)
        engine.validate_model_access(request.model)

        model_class = self._get_model(request.model)
        queryset = model_class.objects.all()

        # Inject scope filters
        injector = ScopeInjector(policy)
        scope_filters = injector.get_scope_filters(request.model, ctx)

        for f in scope_filters:
            queryset = queryset.filter(self._filter_to_q(f))

        # Apply request filters
        if request.where:
            for f in request.where:
                queryset = queryset.filter(self._filter_to_q(f))

        # Store the aggregate operation info
        return CompiledQuery(
            query=queryset,
            request=request,
            injected_filters=scope_filters,
        )

    async def execute_query(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> QueryResult:
        """Execute a compiled query."""
        queryset = compiled.query

        # Execute query and convert to list
        results = list(queryset)

        # Convert model instances to dicts if needed
        data = []
        for item in results:
            if isinstance(item, dict):
                data.append(item)
            else:
                data.append(self._model_to_dict(item, compiled.select_fields))

        return QueryResult(
            data=data,
            total=len(data),
            has_more=False,  # Would need count query for accurate pagination
        )

    async def execute_get(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> GetResult:
        """Execute a compiled get request."""
        queryset = compiled.query
        request = compiled.request

        try:
            result = queryset.first()
            if result is None:
                raise NotFoundError(request.model, request.id)

            if isinstance(result, dict):
                data = result
            else:
                data = self._model_to_dict(result, compiled.select_fields)

            return GetResult(data=data, found=True)
        except Exception as e:
            if isinstance(e, NotFoundError):
                raise
            raise NotFoundError(request.model, request.id) from e

    async def execute_aggregate(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> AggregateResult:
        """Execute a compiled aggregation."""
        queryset = compiled.query
        request = compiled.request

        # Map operation to Django aggregate function
        agg_funcs = {
            "count": Count,
            "sum": Sum,
            "avg": Avg,
            "min": Min,
            "max": Max,
        }

        if request.operation not in agg_funcs:
            raise ValueError(f"Unknown aggregate operation: {request.operation}")

        agg_func = agg_funcs[request.operation]

        if request.operation == "count":
            result = queryset.aggregate(result=Count("*"))
        else:
            if not request.field:
                raise ValueError(f"Field required for {request.operation}")
            result = queryset.aggregate(result=agg_func(request.field))

        return AggregateResult(
            value=result["result"],
            operation=request.operation,
            field=request.field,
        )

    async def transaction(
        self,
        ctx: RunContext,
        fn: Callable[..., T],
        *args: Any,
        **kwargs: Any,
    ) -> T:
        """Execute a function within a transaction."""
        with transaction.atomic():
            return fn(*args, **kwargs)

    def compile_create(
        self,
        request: CreateRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile a create request."""
        engine = PolicyEngine(policy, schema)
        engine.validate_model_access(request.model)
        engine.validate_write_access(request.model, "create")

        model_class = self._get_model(request.model)

        # Add tenant scope to data if configured
        injector = ScopeInjector(policy)
        data = dict(request.data)
        for f in injector.get_scope_filters(request.model, ctx):
            data[f.field] = f.value

        return CompiledQuery(
            query=(model_class, data),
            request=request,
            policy_decisions=["create_allowed"],
        )

    async def execute_create(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> CreateResult:
        """Execute a compiled create request."""
        model_class, data = compiled.query

        instance = model_class.objects.create(**data)

        return CreateResult(
            data=self._model_to_dict(instance),
            id=str(instance.pk),
            success=True,
        )

    def compile_update(
        self,
        request: UpdateRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile an update request."""
        engine = PolicyEngine(policy, schema)
        engine.validate_model_access(request.model)
        engine.validate_write_access(request.model, "update")

        model_class = self._get_model(request.model)
        queryset = model_class.objects.all()

        # Apply scope filters
        injector = ScopeInjector(policy)
        for f in injector.get_scope_filters(request.model, ctx):
            queryset = queryset.filter(self._filter_to_q(f))

        # Filter by ID
        pk_field = self._get_pk_field(model_class)
        queryset = queryset.filter(**{pk_field: request.id})

        return CompiledQuery(
            query=(queryset, request.data),
            request=request,
            policy_decisions=["update_allowed"],
        )

    async def execute_update(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> UpdateResult:
        """Execute a compiled update request."""
        queryset, data = compiled.query
        request = compiled.request

        updated = queryset.update(**data)

        if updated == 0:
            raise NotFoundError(request.model, request.id)

        # Fetch the updated instance
        instance = queryset.first()

        return UpdateResult(
            data=self._model_to_dict(instance) if instance else {},
            updated_count=updated,
            success=True,
        )

    def compile_delete(
        self,
        request: DeleteRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile a delete request."""
        engine = PolicyEngine(policy, schema)
        engine.validate_model_access(request.model)
        engine.validate_write_access(request.model, "delete")

        model_class = self._get_model(request.model)
        queryset = model_class.objects.all()

        # Apply scope filters
        injector = ScopeInjector(policy)
        for f in injector.get_scope_filters(request.model, ctx):
            queryset = queryset.filter(self._filter_to_q(f))

        # Filter by ID
        pk_field = self._get_pk_field(model_class)
        queryset = queryset.filter(**{pk_field: request.id})

        return CompiledQuery(
            query=queryset,
            request=request,
            policy_decisions=["delete_allowed"],
        )

    async def execute_delete(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> DeleteResult:
        """Execute a compiled delete request."""
        queryset = compiled.query
        request = compiled.request

        deleted, _ = queryset.delete()

        if deleted == 0:
            raise NotFoundError(request.model, request.id)

        return DeleteResult(
            deleted_count=deleted,
            success=True,
        )

    def _filter_to_q(self, f: FilterClause) -> Q:
        """Convert a FilterClause to a Django Q object."""
        field = f.field
        value = f.value
        op = f.op

        # Map operators to Django lookups
        lookup_map = {
            "eq": "",
            "ne": "",
            "lt": "__lt",
            "le": "__lte",
            "gt": "__gt",
            "ge": "__gte",
            "in": "__in",
            "nin": "__in",
            "contains": "__icontains",
            "startswith": "__istartswith",
            "endswith": "__iendswith",
            "isnull": "__isnull",
        }

        lookup = lookup_map.get(op, "")
        key = f"{field}{lookup}"

        if op == "ne":
            return ~Q(**{field: value})
        elif op == "nin":
            return ~Q(**{key: value})
        elif op == "isnull":
            return Q(**{key: value})
        else:
            return Q(**{key: value})

    def _get_pk_field(self, model_class: type[models.Model]) -> str:
        """Get the primary key field name for a model."""
        return model_class._meta.pk.name

    def _model_to_dict(
        self,
        instance: models.Model,
        fields: list[str] | None = None,
    ) -> dict[str, Any]:
        """Convert a model instance to a dictionary."""
        data = {}
        field_names = fields if fields else [f.name for f in instance._meta.fields]

        for name in field_names:
            if hasattr(instance, name):
                value = getattr(instance, name)
                # Handle special types
                if hasattr(value, "pk"):  # Foreign key
                    value = value.pk
                data[name] = value

        return data

File: packages/ormai-ts/src/adapters/typeorm/adapter.ts
/**
 * TypeORM adapter for OrmAI.
 */

import type { RunContext } from '../../core/context.js';
import type {
  AggregateRequest,
  AggregateResult,
  BulkUpdateRequest,
  BulkUpdateResult,
  CreateRequest,
  CreateResult,
  DeleteRequest,
  DeleteResult,
  GetRequest,
  GetResult,
  QueryRequest,
  QueryResult,
  UpdateRequest,
  UpdateResult,
} from '../../core/dsl.js';
import type { SchemaMetadata } from '../../core/types.js';
import type { Policy } from '../../policy/models.js';
import { PolicyEngine } from '../../policy/engine.js';
import { PolicyUtils } from '../../policy/models.js';
import type { CompiledMutation, CompiledQuery } from '../base.js';
import { BaseOrmAdapter } from '../base.js';
import {
  TypeORMCompiler,
  type CompiledTypeORMQuery,
  type CompiledTypeORMMutation,
} from './compiler.js';
import { TypeORMIntrospector, type TypeORMDataSource } from './introspection.js';

/**
 * TypeORM repository interface.
 */
export interface TypeORMRepository<T = unknown> {
  find: (options?: TypeORMFindOptions) => Promise<T[]>;
  findOne: (options?: TypeORMFindOptions) => Promise<T | null>;
  save: (entity: T | T[]) => Promise<T | T[]>;
  update: (criteria: unknown, partialEntity: Partial<T>) => Promise<TypeORMUpdateResult>;
  delete: (criteria: unknown) => Promise<TypeORMDeleteResult>;
  createQueryBuilder: (alias?: string) => TypeORMSelectQueryBuilder<T>;
  count: (options?: TypeORMFindOptions) => Promise<number>;
}

/**
 * TypeORM find options.
 */
export interface TypeORMFindOptions {
  where?: Record<string, unknown> | Record<string, unknown>[];
  order?: Record<string, 'ASC' | 'DESC'>;
  take?: number;
  skip?: number;
  select?: string[];
  relations?: string[];
}

/**
 * TypeORM select query builder.
 */
export interface TypeORMSelectQueryBuilder<T = unknown> {
  select: (fields?: string[]) => TypeORMSelectQueryBuilder<T>;
  addSelect: (field: string) => TypeORMSelectQueryBuilder<T>;
  where: (condition: string, parameters?: Record<string, unknown>) => TypeORMSelectQueryBuilder<T>;
  andWhere: (condition: string, parameters?: Record<string, unknown>) => TypeORMSelectQueryBuilder<T>;
  orderBy: (field: string, order: 'ASC' | 'DESC') => TypeORMSelectQueryBuilder<T>;
  addOrderBy: (field: string, order: 'ASC' | 'DESC') => TypeORMSelectQueryBuilder<T>;
  take: (limit: number) => TypeORMSelectQueryBuilder<T>;
  skip: (offset: number) => TypeORMSelectQueryBuilder<T>;
  leftJoinAndSelect: (relation: string, alias: string) => TypeORMSelectQueryBuilder<T>;
  getMany: () => Promise<T[]>;
  getOne: () => Promise<T | null>;
  getCount: () => Promise<number>;
}

/**
 * TypeORM update result.
 */
export interface TypeORMUpdateResult {
  affected?: number;
  raw?: unknown;
}

/**
 * TypeORM delete result.
 */
export interface TypeORMDeleteResult {
  affected?: number;
  raw?: unknown;
}

/**
 * TypeORM entity manager interface.
 */
export interface TypeORMEntityManager {
  getRepository: <T>(entity: string | Function) => TypeORMRepository<T>;
  transaction: <T>(fn: (entityManager: TypeORMEntityManager) => Promise<T>) => Promise<T>;
}

/**
 * Extended TypeORM data source for adapter.
 */
export interface TypeORMAdapterDataSource extends TypeORMDataSource {
  manager: TypeORMEntityManager;
  getRepository: <T>(entity: string | Function) => TypeORMRepository<T>;
  transaction: <T>(fn: (entityManager: TypeORMEntityManager) => Promise<T>) => Promise<T>;
}

/**
 * TypeORM adapter configuration.
 */
export interface TypeORMAdapterConfig {
  /** TypeORM data source */
  dataSource: TypeORMAdapterDataSource;

  /** Model names to expose (optional, defaults to all) */
  models?: string[];
}

/**
 * TypeORM adapter for OrmAI.
 */
export class TypeORMAdapter extends BaseOrmAdapter<
  TypeORMAdapterDataSource,
  CompiledTypeORMQuery,
  CompiledTypeORMMutation
> {
  private readonly dataSource: TypeORMAdapterDataSource;
  private readonly modelFilter?: string[];
  private readonly compiler: TypeORMCompiler;
  private cachedSchema?: SchemaMetadata;

  constructor(config: TypeORMAdapterConfig) {
    super();
    this.dataSource = config.dataSource;
    this.modelFilter = config.models;
    this.compiler = new TypeORMCompiler();
  }

  /**
   * Introspect the TypeORM schema.
   */
  async introspect(): Promise<SchemaMetadata> {
    if (this.cachedSchema) {
      return this.cachedSchema;
    }

    const introspector = new TypeORMIntrospector(this.dataSource, this.modelFilter);
    this.cachedSchema = introspector.introspect();
    return this.cachedSchema;
  }

  /**
   * Get repository for a model.
   */
  private getRepository<T>(
    modelName: string,
    ctx?: RunContext<TypeORMAdapterDataSource>
  ): TypeORMRepository<T> {
    const ds = ctx?.db ?? this.dataSource;
    return ds.getRepository<T>(modelName);
  }

  /**
   * Compile a query request.
   */
  compileQuery(
    request: QueryRequest,
    ctx: RunContext<TypeORMAdapterDataSource>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledQuery<CompiledTypeORMQuery> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateQuery(request, ctx);

    // Merge user filters with scope filters
    const allFilters = [
      ...(request.where ?? []),
      ...decision.injectedFilters,
    ];

    // Compile includes
    const includes = request.include?.map((inc) => inc.relation) ?? [];

    const compiled = this.compiler.compileQuery({
      model: request.model,
      filters: allFilters,
      orderBy: request.orderBy,
      take: request.take,
      select: decision.allowedFields,
      includes,
    });

    return {
      query: compiled,
      request,
      selectFields: decision.allowedFields,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      timeoutMs: decision.budget?.statementTimeoutMs,
    };
  }

  /**
   * Execute a compiled query.
   */
  async executeQuery(
    compiled: CompiledQuery<CompiledTypeORMQuery>,
    ctx: RunContext<TypeORMAdapterDataSource>
  ): Promise<QueryResult> {
    const { query } = compiled;
    const request = compiled.request as QueryRequest;
    const repo = this.getRepository(query.entityName, ctx);

    try {
      let qb = repo.createQueryBuilder(query.alias);

      // Apply select fields
      if (query.selectFields && query.selectFields.length > 0) {
        qb = qb.select(query.selectFields);
      }

      // Apply where conditions
      for (let i = 0; i < query.whereConditions.length; i++) {
        const { condition, parameters } = query.whereConditions[i];
        if (i === 0) {
          qb = qb.where(condition, parameters);
        } else {
          qb = qb.andWhere(condition, parameters);
        }
      }

      // Apply order by
      for (let i = 0; i < query.orderBy.length; i++) {
        const { field, direction } = query.orderBy[i];
        if (i === 0) {
          qb = qb.orderBy(field, direction);
        } else {
          qb = qb.addOrderBy(field, direction);
        }
      }

      // Apply joins
      for (const join of query.joins) {
        qb = qb.leftJoinAndSelect(join.relation, join.alias);
      }

      // Apply pagination
      if (query.take !== undefined) {
        qb = qb.take(query.take);
      }
      if (query.skip !== undefined) {
        qb = qb.skip(query.skip);
      }

      const rows = await qb.getMany();

      // Rows are already filtered to allowedFields via the compiled query
      return {
        data: rows as Record<string, unknown>[],
        nextCursor: null,
        hasMore: rows.length >= request.take,
        totalCount: null,
      };
    } catch (error) {
      throw new Error(`Query execution failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a get request.
   */
  compileGet(
    request: GetRequest,
    ctx: RunContext<TypeORMAdapterDataSource>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledQuery<CompiledTypeORMQuery> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateGet(request, ctx);

    const model = schema.models[request.model];
    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build primary key filter with scope filters
    const pkFilter = { field: model.primaryKey, op: 'eq' as const, value: request.id };
    const allFilters = [pkFilter, ...decision.injectedFilters];

    // Compile includes
    const includes = request.include?.map((inc) => inc.relation) ?? [];

    const compiled = this.compiler.compileQuery({
      model: request.model,
      filters: allFilters,
      select: decision.allowedFields,
      includes,
      take: 1,
    });

    return {
      query: compiled,
      request,
      selectFields: decision.allowedFields,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      timeoutMs: decision.budget?.statementTimeoutMs,
    };
  }

  /**
   * Execute a compiled get query.
   */
  async executeGet(
    compiled: CompiledQuery<CompiledTypeORMQuery>,
    ctx: RunContext<TypeORMAdapterDataSource>
  ): Promise<GetResult> {
    const result = await this.executeQuery(compiled, ctx);
    const row = result.data[0] ?? null;
    return { data: row, found: row !== null };
  }

  /**
   * Compile an aggregate request.
   */
  compileAggregate(
    request: AggregateRequest,
    ctx: RunContext<TypeORMAdapterDataSource>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledQuery<CompiledTypeORMQuery> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateAggregate(request, ctx);

    // Merge user filters with scope filters
    const allFilters = [
      ...(request.where ?? []),
      ...decision.injectedFilters,
    ];

    const compiled = this.compiler.compileQuery({
      model: request.model,
      filters: allFilters,
    });

    return {
      query: compiled,
      request,
      selectFields: [],
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      timeoutMs: decision.budget?.statementTimeoutMs,
    };
  }

  /**
   * Execute a compiled aggregate query.
   */
  async executeAggregate(
    compiled: CompiledQuery<CompiledTypeORMQuery>,
    ctx: RunContext<TypeORMAdapterDataSource>
  ): Promise<AggregateResult> {
    const { query } = compiled;
    const request = compiled.request as AggregateRequest;
    const repo = this.getRepository(query.entityName, ctx);

    try {
      let qb = repo.createQueryBuilder(query.alias);

      // Apply where conditions
      for (let i = 0; i < query.whereConditions.length; i++) {
        const { condition, parameters } = query.whereConditions[i];
        if (i === 0) {
          qb = qb.where(condition, parameters);
        } else {
          qb = qb.andWhere(condition, parameters);
        }
      }

      let value: unknown = null;
      let rowCount = 0;

      switch (request.operation) {
        case 'count':
          value = await qb.getCount();
          rowCount = value as number;
          break;
        default: {
          // For other aggregates, fetch rows and compute client-side
          const rows = await qb.getMany() as Record<string, unknown>[];
          rowCount = rows.length;

          if (request.field) {
            switch (request.operation) {
              case 'sum':
                value = rows.reduce((sum, row) => sum + (Number(row[request.field!]) || 0), 0);
                break;
              case 'avg':
                if (rows.length > 0) {
                  value = rows.reduce((sum, row) => sum + (Number(row[request.field!]) || 0), 0) / rows.length;
                }
                break;
              case 'min':
                if (rows.length > 0) {
                  value = Math.min(...rows.map((row) => Number(row[request.field!]) || 0));
                }
                break;
              case 'max':
                if (rows.length > 0) {
                  value = Math.max(...rows.map((row) => Number(row[request.field!]) || 0));
                }
                break;
            }
          }
        }
      }

      return {
        value,
        operation: request.operation,
        field: request.field ?? null,
        rowCount,
      };
    } catch (error) {
      throw new Error(`Aggregate execution failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a create request.
   */
  compileCreate(
    request: CreateRequest,
    ctx: RunContext<TypeORMAdapterDataSource>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledTypeORMMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateCreate(request, ctx);

    // Build data with scope injection
    const data = { ...request.data };
    for (const filter of decision.injectedFilters) {
      if (filter.op === 'eq') {
        data[filter.field] = filter.value;
      }
    }

    const compiled = this.compiler.compileInsert({
      model: request.model,
      data,
    });

    return {
      mutation: compiled,
      request,
      data,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: decision.allowedFields,
    };
  }

  /**
   * Execute a compiled create mutation.
   */
  async executeCreate(
    compiled: CompiledMutation<CompiledTypeORMMutation>,
    ctx: RunContext<TypeORMAdapterDataSource>
  ): Promise<CreateResult> {
    const { mutation } = compiled;
    const repo = this.getRepository(mutation.entityName, ctx);

    try {
      const result = await repo.save(mutation.data as Record<string, unknown>);

      return {
        data: result as Record<string, unknown>,
        id: (result as Record<string, unknown>)['id'],
        success: true,
      };
    } catch (error) {
      throw new Error(`Create failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile an update request.
   */
  compileUpdate(
    request: UpdateRequest,
    ctx: RunContext<TypeORMAdapterDataSource>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledTypeORMMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateUpdate(request, ctx);

    const model = schema.models[request.model];
    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build primary key filter with scope filters
    const pkFilter = { field: model.primaryKey, op: 'eq' as const, value: request.id };
    const allFilters = [pkFilter, ...decision.injectedFilters];

    const compiled = this.compiler.compileUpdate({
      model: request.model,
      data: request.data,
      filters: allFilters,
    });

    return {
      mutation: compiled,
      request,
      data: request.data,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: decision.allowedFields,
    };
  }

  /**
   * Execute a compiled update mutation.
   */
  async executeUpdate(
    compiled: CompiledMutation<CompiledTypeORMMutation>,
    ctx: RunContext<TypeORMAdapterDataSource>
  ): Promise<UpdateResult> {
    const { mutation } = compiled;
    const request = compiled.request as UpdateRequest;
    const repo = this.getRepository(mutation.entityName, ctx);
    const schema = await this.introspect();
    const model = schema.models[mutation.entityName];
    const pkField = model.primaryKey;

    try {
      const result = await repo.update(
        { [pkField]: request.id } as unknown,
        mutation.data as Record<string, unknown>
      );

      // Fetch the updated record
      const updated = await repo.findOne({
        where: { [pkField]: request.id } as Record<string, unknown>,
      });

      return {
        data: updated as Record<string, unknown> | null,
        success: (result.affected ?? 0) > 0,
        found: (result.affected ?? 0) > 0,
      };
    } catch (error) {
      throw new Error(`Update failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a delete request.
   */
  compileDelete(
    request: DeleteRequest,
    ctx: RunContext<TypeORMAdapterDataSource>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledTypeORMMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateDelete(request, ctx);

    const model = schema.models[request.model];
    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build primary key filter with scope filters
    const pkFilter = { field: model.primaryKey, op: 'eq' as const, value: request.id };
    const allFilters = [pkFilter, ...decision.injectedFilters];

    const rowPolicy = PolicyUtils.getRowPolicy(policy, request.model);
    const softDelete = !request.hard && rowPolicy.softDeleteField;

    if (softDelete) {
      // Soft delete is an update
      const compiled = this.compiler.compileUpdate({
        model: request.model,
        data: { [rowPolicy.softDeleteField!]: new Date() },
        filters: allFilters,
      });

      return {
        mutation: compiled,
        request,
        data: { [rowPolicy.softDeleteField!]: new Date() },
        injectedFilters: decision.injectedFilters,
        policyDecisions: [...decision.decisions, 'Using soft delete'],
        returnFields: [],
      };
    }

    const compiled = this.compiler.compileDelete({
      model: request.model,
      filters: allFilters,
    });

    return {
      mutation: compiled,
      request,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: [],
    };
  }

  /**
   * Execute a compiled delete mutation.
   */
  async executeDelete(
    compiled: CompiledMutation<CompiledTypeORMMutation>,
    ctx: RunContext<TypeORMAdapterDataSource>
  ): Promise<DeleteResult> {
    const { mutation } = compiled;
    const request = compiled.request as DeleteRequest;
    const repo = this.getRepository(mutation.entityName, ctx);
    const schema = await this.introspect();
    const model = schema.models[mutation.entityName];
    const pkField = model.primaryKey;

    try {
      if (mutation.type === 'update') {
        // Soft delete
        const result = await repo.update(
          { [pkField]: request.id } as unknown,
          mutation.data as Record<string, unknown>
        );

        return {
          success: (result.affected ?? 0) > 0,
          found: (result.affected ?? 0) > 0,
          softDeleted: true,
        };
      } else {
        // Hard delete
        const result = await repo.delete({ [pkField]: request.id } as unknown);

        return {
          success: (result.affected ?? 0) > 0,
          found: (result.affected ?? 0) > 0,
          softDeleted: false,
        };
      }
    } catch (error) {
      throw new Error(`Delete failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a bulk update request.
   */
  compileBulkUpdate(
    request: BulkUpdateRequest,
    ctx: RunContext<TypeORMAdapterDataSource>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledTypeORMMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateBulkUpdate(request, ctx);

    const model = schema.models[request.model];
    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build where clause with all IDs and scope filters
    const idsFilter = { field: model.primaryKey, op: 'in' as const, value: request.ids };
    const allFilters = [idsFilter, ...decision.injectedFilters];

    const compiled = this.compiler.compileUpdate({
      model: request.model,
      data: request.data,
      filters: allFilters,
    });

    return {
      mutation: compiled,
      request,
      data: request.data,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: [],
    };
  }

  /**
   * Execute a compiled bulk update mutation.
   */
  async executeBulkUpdate(
    compiled: CompiledMutation<CompiledTypeORMMutation>,
    ctx: RunContext<TypeORMAdapterDataSource>
  ): Promise<BulkUpdateResult> {
    const { mutation } = compiled;
    const repo = this.getRepository(mutation.entityName, ctx);

    try {
      // Build criteria from where conditions
      const criteria: Record<string, unknown> = {};
      for (const { parameters } of mutation.whereConditions) {
        Object.assign(criteria, parameters);
      }

      const result = await repo.update(
        criteria,
        mutation.data as Record<string, unknown>
      );

      return {
        updatedCount: result.affected ?? 0,
        success: (result.affected ?? 0) > 0,
        failedIds: [],
      };
    } catch (error) {
      throw new Error(`Bulk update failed: ${(error as Error).message}`);
    }
  }

  /**
   * Execute operations in a transaction.
   */
  async transaction<T>(
    ctx: RunContext<TypeORMAdapterDataSource>,
    fn: () => Promise<T>
  ): Promise<T> {
    const ds = ctx.db ?? this.dataSource;
    return ds.transaction(async () => fn());
  }
}

/**
 * Create a TypeORM adapter.
 */
export function createTypeORMAdapter(config: TypeORMAdapterConfig): TypeORMAdapter {
  return new TypeORMAdapter(config);
}
